common/dm.c

```diff
@@ -0,0 +1,189 @@
+#include <libgen.h>
+#include <stdlib.h>
+#include <sys/mount.h>
+#include <errno.h>
+#include <stdint.h>
+
+#include "macro.h"
+#include "mem.h"
+#include "dm.h"
+
+#define ARRAY_SIZE(x) (sizeof(x) / sizeof((x)[0]))
+
+struct dm_cmd_table cmd_table[] = {
+	{ DM_DEV_CREATE, { 4, 0, 0 } },	  { DM_TABLE_LOAD, { 4, 0, 0 } },
+	{ DM_DEV_REMOVE, { 4, 0, 0 } },	  { DM_REMOVE_ALL, { 4, 0, 0 } },
+	{ DM_DEV_SUSPEND, { 4, 0, 0 } },  { DM_DEV_STATUS, { 4, 0, 0 } },
+	{ DM_TABLE_DEPS, { 4, 0, 0 } },	  { DM_DEV_RENAME, { 4, 0, 0 } },
+	{ DM_VERSION, { 4, 0, 0 } },	  { DM_TABLE_STATUS, { 4, 0, 0 } },
+	{ DM_DEV_WAIT, { 4, 0, 0 } },	  { DM_LIST_DEVICES, { 4, 0, 0 } },
+	{ DM_TABLE_CLEAR, { 4, 0, 0 } },  { DM_LIST_VERSIONS, { 4, 1, 0 } },
+	{ DM_TARGET_MSG, { 4, 2, 0 } },	  { DM_DEV_SET_GEOMETRY, { 4, 6, 0 } },
+	{ DM_DEV_ARM_POLL, { 4, 36, 0 } }
+};
+
+#ifndef __GNU_LIBRARY__
+/*
+ * non glibc std libraries such as musl provide ioctl (int, int, ...)
+ * wrapper. However, dm integrity requests are 'unsigned long int' and
+ * would overflow on a cast to int. Thus, we directly provide a wrapper
+ * here instead of using the ioctl wrapper of the std library.
+ */
+int
+dm_ioctl(int fd, unsigned long int request, ...)
+{
+	void *args;
+	va_list ap;
+	int result;
+	va_start(ap, request);
+	args = va_arg(ap, void *);
+	result = syscall(__NR_ioctl, fd, request, args);
+	va_end(ap);
+	return result;
+}
+#endif
+
+int
+dm_ioctl_init(struct dm_ioctl *io, enum dm_cmd_index idx, size_t data_size, const char *name,
+	      const char *uuid, unsigned flags, unsigned long long dev, unsigned int target_count,
+	      unsigned int event_nr)
+{
+	if (idx > ARRAY_SIZE(cmd_table)) {
+		ERROR("Failed to lookup ioctl command");
+		return -1;
+	}
+
+	mem_memset(io, 0, data_size);
+	io->data_size = data_size;
+	io->data_start = sizeof(struct dm_ioctl);
+	io->version[0] = cmd_table[idx].version[0];
+	io->version[1] = cmd_table[idx].version[1];
+	io->version[2] = cmd_table[idx].version[2];
+	io->flags = flags;
+	io->dev = dev;
+	io->target_count = target_count;
+	io->event_nr = event_nr;
+
+	if (name)
+		strncpy(io->name, name, sizeof(io->name) - 1);
+
+	if (uuid)
+		strncpy(io->uuid, uuid, sizeof(io->name) - 1);
+
+	return 0;
+}
+
+uint64_t
+dm_get_blkdev_size64(int fd)
+{
+	uint64_t size;
+
+	if ((dm_ioctl(fd, BLKGETSIZE64, &size)) == -1) {
+		ERROR_ERRNO("BLKGETSIZE64 ioctl failed");
+		size = 0;
+	}
+	return size;
+}
+
+int
+dm_get_blkdev_sector_size(int fd)
+{
+	int sec_size;
+
+	if ((dm_ioctl(fd, BLKSSZGET, &sec_size)) == -1) {
+		ERROR_ERRNO("BLKSSZGET ioctl failed");
+		sec_size = -1;
+	}
+	return sec_size;
+}
+
+int
+dm_get_blkdev_readonly(int fd)
+{
+	int read_only;
+
+	if ((dm_ioctl(fd, BLKROGET, &read_only)) == -1) {
+		ERROR_ERRNO("BLKROGET ioctl failed");
+		read_only = -1;
+	}
+	return read_only;
+}
+
+int
+dm_read_version(void)
+{
+	int control_fd = -1;
+	uint8_t buf[16384] = { 0 };
+	struct dm_ioctl *dmi = NULL;
+
+	control_fd = open("/dev/mapper/control", O_RDWR);
```

Comment:
use define. in cryptfs.c is a define block which could be moved to dm.h:
#ifdef ANDROID
#define DEV_MAPPER "/dev/device-mapper"
#define CRYPT_PATH_PREFIX "/dev/block/dm-"
#else
#define DEV_MAPPER "/dev/mapper/control"
#define CRYPT_PATH_PREFIX "/dev/mapper/"
#endif

Reply:
DEV_MAPPER -> DM_DEV_MAPPER
CRYPT_PATH_PREFIX -> DM_PATH_PREFIX

Reply:
Done